lib: lwm2m_carrier: add liblwm2m_carrier

lemrey opened the pull request:
Add the LwM2M carrier library under `/lib/lwm2m_carrier/`,
one dedicated sample and an overlay for the `mqtt_simple`, `coap_client` projects.

Depends on #1122.

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:169: WARNING:NEW_TYPEDEFS: do not add new typedefs
#169: FILE: lib/lwm2m_carrier/include/lwm2m_carrier.h:26:
+typedef struct {

-:171: WARNING:PREFER_KERNEL_TYPES: Prefer kernel type 'u32_t' over 'uint32_t'
#171: FILE: lib/lwm2m_carrier/include/lwm2m_carrier.h:28:
+	uint32_t type;

-:177: WARNING:TYPO_SPELLING: 'libary' may be misspelled - perhaps 'library'?
#177: FILE: lib/lwm2m_carrier/include/lwm2m_carrier.h:34:
+ * @brief Structure holding LWM2M carrier libary initialization parameters.

-:179: WARNING:NEW_TYPEDEFS: do not add new typedefs
#179: FILE: lib/lwm2m_carrier/include/lwm2m_carrier.h:36:
+typedef struct {

-:222: WARNING:PREFER_KERNEL_TYPES: Prefer kernel type 's32_t' over 'int32_t'
#222: FILE: lib/lwm2m_carrier/include/lwm2m_carrier.h:79:
+const char *lwm2m_carrier_timezone(int32_t tz_offset, int32_t dst);

-:232: WARNING:TYPO_SPELLING: 'occured' may be misspelled - perhaps 'occurred'?
#232: FILE: lib/lwm2m_carrier/include/lwm2m_carrier.h:89:
+ * @param[in] event A LWM2M carrier event that occured.

-:313: WARNING:PREFER_KERNEL_TYPES: Prefer kernel type 'u16_t' over 'uint16_t'
#313: FILE: lib/lwm2m_carrier/include/lwm2m_os.h:71:
+int lwm2m_os_storage_read(uint16_t id, void *data, size_t len);

-:318: WARNING:PREFER_KERNEL_TYPES: Prefer kernel type 'u16_t' over 'uint16_t'
#318: FILE: lib/lwm2m_carrier/include/lwm2m_os.h:76:
+int lwm2m_os_storage_write(uint16_t id, const void *data, size_t len);

-:481: WARNING:TYPO_SPELLING: 'mangement' may be misspelled - perhaps 'management'?
#481: FILE: lib/lwm2m_carrier/os/lwm2m_os.c:44:
+/* Memory mangement. */

-:525: WARNING:PREFER_KERNEL_TYPES: Prefer kernel type 'u16_t' over 'uint16_t'
#525: FILE: lib/lwm2m_carrier/os/lwm2m_os.c:88:
+int lwm2m_os_storage_read(uint16_t id, void *data, size_t len)

-:530: WARNING:PREFER_KERNEL_TYPES: Prefer kernel type 'u16_t' over 'uint16_t'
#530: FILE: lib/lwm2m_carrier/os/lwm2m_os.c:93:
+int lwm2m_os_storage_write(uint16_t id, const void *data, size_t len)

-:750: ERROR:POINTER_LOCATION: "foo * bar" should be "foo *bar"
#750: FILE: samples/nrf9160/coap_client/src/main.c:189:
+void lwm2m_carrier_event_handler(const lwm2m_carrier_event_t * event)

-:1026: ERROR:POINTER_LOCATION: "foo * bar" should be "foo *bar"
#1026: FILE: samples/nrf9160/mqtt_simple/src/main.c:57:
+void lwm2m_carrier_event_handler(const lwm2m_carrier_event_t * event)

- total: 2 errors, 11 warnings, 953 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.



```
## Codeowners issues

```
New files added that are not covered in CODEOWNERS:

lib/lwm2m_carrier/CHANGELOG.rst
lib/lwm2m_carrier/CMakeLists.txt
lib/lwm2m_carrier/include/lwm2m_carrier.h
lib/lwm2m_carrier/include/lwm2m_os.h
lib/lwm2m_carrier/lib/cortex-m33/hard-float/liblwm2m_carrier.a
lib/lwm2m_carrier/os/bootstrap_psk.h
lib/lwm2m_carrier/os/lwm2m_carrier.c
lib/lwm2m_carrier/os/lwm2m_os.c

Please add one or more entries in the CODEOWNERS file to cover those files


```

stig-bjorlykke commented on lib/lwm2m_carrier/include/lwm2m_os.h:
```diff
@@ -0,0 +1,113 @@
+#include <stdint.h>
+
+/*
+ * @brief Maximum number of timers that system must support.
```
This should probably be "Minimum number of timers that the system must support."?

lemrey replied:
I think that this is actually correct, that's the maximum number of times that the OS must support in order to run this library.

metc commented on lib/lwm2m_carrier/CHANGELOG.rst:
```diff
@@ -0,0 +1,11 @@
+All notable changes to this project are documented in this file.
+
+liblwm2m_carrier 0.6.0
+***************
```
I think you need to add some padding here to match the zephyr style

lemrey replied:
Done

metc commented on samples/nrf9160/lwm2m_carrier/src/main.c:
```diff
@@ -0,0 +1,47 @@
+
+void bsd_recoverable_error_handler(uint32_t error)
+{
+	printk("RECOVERABLE ERROR %lu\n", (unsigned long)error);
```
All examples should be aligned. Printing an unsigned `%u` value should be enough here.

lemrey replied:
Done.

lemrey commented on samples/nrf9160/lwm2m_carrier/prj.conf:
```diff
@@ -0,0 +1,50 @@
+CONFIG_LWM2M_CARRIER=y
+
+# LWM2M Carrier requires newlib
+CONFIG_NEWLIB_LIBC=y
```
Me and Robert have discussed this and basically there are many limitations to `select` in Kconfig, i.e. we would only be able to turn on a setting. It would be impossible to turn something off or change a numerical/string value. Thus, it might be better not to scatter the required Kconfig settings between Kconfig/`prj.conf` and keep them in one place instead.

lemrey replied:
That is not a bad idea, it is possible to `depend on` all these settings indeed.
However even if all dependencies were in Kconfig, turning `LWM2M_CARRIER` on won't select any of them; we would still have to provide a `prj.conf` file that shows what the library needs, and then that information would be duplicated in both `prj.conf` and Kconfig. It is true that having the dependencies in Kconfig prevents the user from configuring it incorrectly but the price is having to duplicate the list of dependencies. I don't see a really big problem with that, just a very big Kconfig file :)

SebastianBoe replied:
This price is acceptable IMHO, assuming the user is supposed to be able to add the LWM2M carrier to his own application and not just use the lwm2m_carrier app.

lemrey replied:
Added dependencies to Kconfig.

lemrey commented on samples/nrf9160/lwm2m_carrier/prj.conf:
```diff
@@ -0,0 +1,50 @@
+#
+
+CONFIG_LWM2M_CARRIER=y
+
```
Correct, will fix.

lemrey replied:
Fixed.

lemrey replied:
Done

SebastianBoe commented on lib/lwm2m_carrier/Kconfig:
```diff
@@ -0,0 +1,26 @@
+	help
+	  A library for cellular connection management.
+
+if (LWM2M_CARRIER)
```
parenthesis should be dropped

lemrey replied:
Remove parenthesis.

SebastianBoe commented on lib/lwm2m_carrier/os/lwm2m_os.c:
```diff
@@ -0,0 +1,235 @@
+	int err;
+
+	/* Initialize storage */
+	err = nvs_init(&fs, DT_FLASH_DEV_NAME);
```
Just

`return nvs_init(&fs, DT_FLASH_DEV_NAME);
`

does the same and is easier to read.

lemrey replied:
True, we had more stuff in that function so that's why it looks like that :)
I think it's fine to keep it, in case more things will need to happen during initialization.

SebastianBoe replied:
I disagree, it can be expanded at that time.

Right now it's communicating that it's doing more than it is.

SebastianBoe commented on lib/lwm2m_carrier/os/lwm2m_os.c:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	/* Zephyr's timing subsystem uses positive integers so the
+	 * largest tick count that can be represented is 31 bit large.
```
Shouldn't this be saying:

> Zephyr's timing subsystem uses **signed** integers so the

?

lemrey replied:
You are very correct, fixed.

SebastianBoe commented on lib/lwm2m_carrier/os/lwm2m_os.c:
```diff
@@ -0,0 +1,235 @@
+	timeout = MIN(K_SECONDS(max_timeout_sec), timeout);
+
+	/* FIXME: return an error if the timeout requested is too large. */
+	return k_delayed_work_submit(&work->work_item, K_MSEC(timeout));
```
I don't get it.

Both the wrapper, and the wrappee are expressing timeouts in signed 32 bit milliseconds, so why do we need to do any correction?

lemrey replied:
Internally the timeout is converted in number of ticks and those are signed as far I understand.

SebastianBoe replied:
> Internally the timeout is converted in number of ticks

What is working with ticks? **k_delayed_work_submit** definitely works with milliseconds, and AFAICT **lwm2m_os_timer_start** does as well, or else you would get the wrong behaviour when timeout == 1 for instance.

SebastianBoe replied:
My suggestion is to delete line 171 and line 165.

lemrey replied:
See [here](https://github.com/zephyrproject-rtos/zephyr/blob/82c25587a2bc7610530154df4d984bdfa2add594/kernel/work_q.c#L110). Notice:
- `void z_add_timeout(struct _timeout *to, _timeout_func_t fn, s32_t ticks)` and
- ` s32_t z_ms_to_ticks(s32_t ms)` signatures.

In particular tick counting is signed.

SebastianBoe replied:
Right, k_delayed_work_submit uses ticks internally, but externally both k_delayed_work_submit and lwm2m_os_timer_start uses ms, so we don't need to touch the timeout value in this function.

lemrey replied:
The code here detects if any timeout in milliseconds can be expressed correctly in ticks before submitting the work item, it's not doing a conversion. When a timeout of 1 is requested, it will timeout in one millisecond.
Note that: `#define K_MSEC(ms)     (ms)`,
I have just added that to make the unit very explicit.

SebastianBoe replied:
Isn't it the responsiblity of **k_delayed_work_submit** to detect this? How is this a unique requirement of **lwm2m_os_timer_start**?

What happens when it can't be expressed and why can't it always be expressed?

lemrey replied:
This is a limitation of Zephyr's current timing system, as far as I understand..
If a timeout that becomes negative when translated into ticks (`int`) is treated as if it had expired and the delayed work will be fired off at the earliest.

For example: `K_HOURS(20)` , would be translated to ticks as:
`20 * 60 * 60 * CONFIG_SYS_CLOCK_TICKS_PER_SEC (32768) = 0x8CA00000`.
That number is negative and submitting a work item with a delay of `K_HOURS(20)` will make the work item expire immediately instead of in twenty hours.

SebastianBoe replied:
This should be reported as a bug and a reference to the bugreport should be listed next to our workaround code.

lemrey replied:
Perhaps I can could a reference to this: https://github.com/zephyrproject-rtos/zephyr/issues/9663? Otherwise I could create a new issue and reference that, but I think there are already many issues about this topic.

SebastianBoe replied:
That is not relevant. A new issue should be created, it will be closed as duplicate if appropriate.

lemrey replied:
Done.

SebastianBoe commented on .gitignore:
```diff
@@ -1,5 +1,4 @@
 *.o
-*.a
```
If there is a .a file lying around in my repo, then I would want to know about it when I run git status.

I don't see why one would want to gitignore .a files.

lemrey replied:
So you are okay with me removing it from .gitignore?

SebastianBoe replied:
Yes

SebastianBoe replied:
Oh, sorry about that, didn't see the red color.

SebastianBoe commented on lib/lwm2m_carrier/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+#
+
+if(CONFIG_FLOAT)
+	if(CONFIG_FP_HARDABI)
```
Indentation is two spaces.

lemrey replied:
Fixed.

SebastianBoe commented on lib/lwm2m_carrier/include/lwm2m_os.h:
```diff
@@ -0,0 +1,113 @@
+/**
+ * @brief Log a message.
+ */
+void lwm2m_os_log(int level, const char *fmt, ...);
```
I assume liblwm2m_carrier.a links with this API. Doesn't that then make it an RTOS-independent library belonging in nrfxlib?

lemrey replied:
It has other dependencies on NCS-only libraries.

SebastianBoe commented on lib/lwm2m_carrier/os/lwm2m_carrier.c:
```diff
@@ -0,0 +1,53 @@
+	}
+
+	/* Non-return function. */
+	lwm2m_carrier_run();
```
Drop the comment and put

	CODE_UNREACHABLE;

after lwm2m_carrier_run();

Then both the compiler and the reader will know that it is a "Non-return function".

lemrey replied:
Right, done.

lemrey commented on samples/nrf9160/lwm2m_carrier/prj.conf:
```diff
@@ -0,0 +1,50 @@
+# LTE link control
+CONFIG_LTE_LINK_CONTROL=y
+# Disable autoconnect
+CONFIG_LTE_AUTO_INIT_AND_CONNECT=n
```
Added dependencies to Kconfig.

lemrey commented:
I have added myself and @rlubos to `CODEOWNERS` for the library and squashed the documentation commits. I should have addressed all comments so far.

lemrey commented:
Rebased in an effort to pass CI.
